dockerfiles/haskell/Dockerfile

```diff
@@ -0,0 +1,29 @@
+FROM golang:1.10-alpine
+WORKDIR /go/src/github.com/sourcegraph/lsp-adapter
+COPY . .
+RUN CGO_ENABLED=0 GOBIN=/usr/local/bin go install github.com/sourcegraph/lsp-adapter
+
+FROM ubuntu:16.04
+RUN apt-get update \
+  # libtinfo-dev is required by haskell-ide-engine.
+  && apt-get install -y git wget libtinfo-dev \
+  && wget -qO- https://get.haskellstack.org/ | sh \
+  && git clone https://github.com/haskell/haskell-ide-engine --recursive \
```

Comment:
Is it possible to download a zip instead? Otherwise this would always download the full history, which bloats the image but also invalidates the cache on every new commit even though we just want a specific one.

Reply:
Yes, but it's 57 MB, which is tiny compared to the final image which is 5 GB 😱 

How would it invalidate the cache? The RUN command won't change.

> unless the RUN command *itself* changes (and thus invalidates Docker’s on-host cache), Docker will reuse the previous results from cache. 

https://thenewstack.io/understanding-the-docker-cache-for-faster-builds/

Reply:
So if I have a RUN command like `npm install my-package` and `my-package` releases a new version, I rebuild the Docker image and it installs the new version, it won't actually use the new version because the `RUN` command didn't change? That doesn't make sense to me - if that was true there would be no point in running the command at all if it didn't change and Docker could just fast exit if nothing about a Dockerfile changed. If the file system changed after a RUN, that new state must make it into the image and result in a new hash, invalidating the cache.

Size really isn't my concern here, but we build all images on every commit so we should take every measure we can to not publish hundreds of "noop" versions.

Reply:
I did some [research](https://windsock.io/explaining-docker-image-ids/) 📖  and experimentation 🔬 

You're right: `git clone` will result in an image with a different set of layers, causing a lot of layer churn, so I'll change this to a zip file :+1:

The RUN cache that I was talking about only caches intermediate build steps on the local machine, but doesn't help when CI builds the whole image from the base on each commit.

We already have a zillion tags on Docker Hub (e.g. https://hub.docker.com/r/sourcegraph/codeintel-elixir/tags/). Every commit will give each image another tag. Maybe we can figure out how to avoid tagging a new image if the digest didn't change, but I don't think it's particularly harmful since it's not the image ID or tag that matter when downloading but instead the layer digests.

Reply:
Shucks, the zip file from GitHub doesn't contain the contents of submodules. I'll `rm -rf` the `.git` directories instead.

Reply:
Imo if tags are a problem we should consider not tagging at all, but tags are only useful if they exist for every commit.

Reply:
@felixfbecker 

> it won't actually use the new version because the RUN command didn't change

Correct. All docker knows about is the command you run. If the arguments to the command have not changed, and if a previous build phase didn't already bust the cache, then it will use a cached image. 

Read https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#build-cache

Reply:
> Correct. All docker knows about is the command you run. If the arguments to the command have not changed, and if a previous build phase didn't already bust the cache, then it will use a cached image.

Where would that build cache come from in CI?

Reply:
The build machine itself that is running the docker engine presumably.

Reply:
Not sure I understand, every CI job is a fresh container instance - wouldn't this require persisting that Docker build cache from previous CI build?

Reply:
https://travis-ci.org/sourcegraph/lsp-adapter/jobs/378676930

It looks like nothing hits the cache, including the base image, which seems to be evidence that builds are completely isolated.

My philosophy here is to just assume that a cached is possible (since Docker is designed this way) and to write Dockerfiles accordingly so it doesn't matter where they are run.

Reply:
> to write Dockerfiles accordingly so it doesn't matter where they are run.

I agree with the philosophy, but in this case assuming a cache is always hit _makes_ it matter where they run as the result will always be different in CI. Deleting changing/not needed files seems to be best practice in Docker (deleting apt lists, deleting .git, etc) so I would propose to do as @chrismwendt suggested